### ISSUE TITLE
Kong sometimes doesn't use port from SRV lookup

### ISSUE THREAD

**Issue by nvx** [Kong/kong#7786]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

2.5.0

### Current Behavior

When using an upstream URI of `http://example-query.query.consul` with Kong configured with `dns_order = SRV,A,CNAME` and the resolver pointing to Consul, occasionally in the logs it looks like Kong is ignoring the port from the SRV lookup.

Logs (via the udp-log) plugin shows tries of x.x.x.130:80, x.x.x.114:80, x.x.x.130:80, x.x.x.114:80, x.x.x.130:80, x.x.x.114:80, while doing a DNS lookup of that hostname shows the same IPs, but with ports:

```
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;example-query.query.consul. IN SRV

example-query.query.consul. 0 IN SRV 1 1 24866 0a143372.addr.tsv-dc.consul.
example-query.query.consul. 0 IN SRV 1 1 30960 0a143382.addr.tsv-dc.consul.

;; ADDITIONAL SECTION:
xxxxxx72.addr.tsv-dc.consul. 0 IN A  x.x.x.114
ei-tsv-prod-nomad-green-2.node.tsv-dc.consul. 0 IN TXT "consul-network-segment="
xxxxxx82.addr.tsv-dc.consul. 0 IN A  x.x.x.130
ei-tsv-prod-nomad-green-1.node.tsv-dc.consul. 0 IN TXT "consul-network-segment="
```

Note LAST is not in the dns_order, so Kong should always prefer a SRV lookup. It is possible that the SRV lookup failed (although this should be unlikely as the Consul agent services requests locally) and it retried with an A lookup (which wouldn't have had port information), but in that case I'd have expected only the first try to be port 80 then subsequent attempts to retry with a port, or is the DNS resolution not repeated for subsequent attempts? Also odd is sometimes there can be multiple different requests to the same service in a row all failing in the same fashion, is Kong doing any caching of DNS responses (ie, caching that the SRV lookup failed) that would be causing it to fall back to an A lookup?

If it is failing the SRV lookup and falling back to A, is there any way to force some lookups only ever be SRV while still permitting others to be A lookups (ie, a flag per-service or *.consul. FQDNs)? I encountered a similar issue a long time ago which caused me to remove LAST from the dns_order list where short term issues (like the service being down) resulted in LAST being remembered as A (because the service was back up by that stage), but unfortunately I can't remove A records entirely from being looked up as some still require that.

### Expected Behavior

Expected Kong to respect the port from the SRV lookup

### Steps To Reproduce

```markdown
Intermittent issue
```

**Comment by nvx** [comment]
I've noticed the dns_error_ttl and dns_not_found_ttl options which might explain some of the behaviour I'm seeing, but I'm not 100% sure if it applies in this case (ie, does it remember if a SRV lookup failed separately from an A lookup, or if it's just cached on the hostname). If it does apply reducing those could minimise the issue, but understanding why it failed over to doing an A lookup would be useful, and will setting one or more of these to 0 result in Kong re-attempting the DNS lookup for each try?

**Comment by locao** [comment]
Yes, Kong caches the DNS responses, and it respects the record's TTL.

The port `80` is the default when no port is configured in the target entity and the entry was not resolved to an SRV record. Can you try the debugging suggestion in [this comment](https://github.com/Kong/kong/issues/7771#issuecomment-914313390), please? This way we can check if the resolver is always working.

**Comment by nvx** [comment]
> Yes, Kong caches the DNS responses, and it respects the record's TTL.
> 
> The port `80` is the default when no port is configured in the target entity and the entry was not resolved to an SRV record. Can you try the debugging suggestion in [this comment](https://github.com/Kong/kong/issues/7771#issuecomment-914313390), please? This way we can check if the resolver is always working.

I'm not sure why I didn't think of this earlier *facepalm*, rather than tcpdump I instead enabled query logging in unbound (which runs on the same server and is queried over the loopback interface) since the output is nicer. Sure enough I can see a bunch of SRV lookups with NOERROR being returned, then just before the fails appear I get a couple of SERVFAIL's (I'm guessing one for each nginx worker process) so looks like my root cause troubleshooting moves upstream.

What I am curious about now though in an attempt to mitigate the upstream issue is the difference between these config options:

```
dns_not_found_ttl = 1            # TTL in seconds for empty DNS responses and
                                 # "(3) name error" responses.

dns_error_ttl = 1                # TTL in seconds for error responses.
```

Is SERVFAIL cached for dns_error_ttl duration? Or is SERVFAIL considered a "(3) name error" (I couldn't work out what name error is meant to be, is it NXDOMAIN which has a DNS RCODE of 3?)

And is it safe to set eg dns_error_ttl to 0 to prevent caching of "SRV failed" and instead have it retry each time?

**Comment by Tieske** [comment]
> I couldn't work out what name error is meant to be, is it NXDOMAIN which has a DNS RCODE of 3?

yes.

so:
- success; `NOERROR` is cached by the records TTL
- empty responses as well as `NXDOMAIN` (RCODE 3) get cached by `dns_not_found_ttl`
- every other return code gets cached by `dns_error_ttl`

**Comment by Tieske** [comment]
I actually found a bug yesterday (I believe) if a name has both a SRV and an A record, then querying (with SRV first) will not return the SRV port. The test record I use is `srvrecurse.thijsschreijer.nl`.

(Kong internal tracking number; FTI-2800)

**Comment by nvx** [comment]
> I actually found a bug yesterday (I believe) if a name has both a SRV and an A record, then querying (with SRV first) will not return the SRV port. The test record I use is `srvrecurse.thijsschreijer.nl`.
> 
> (Kong internal tracking number; FTI-2800)

Interesting. I was able to resolve my specific issue with some tweaks to the configuration in unbound (increasing the server.infra-cache-min-rtt parameter to "1000" (ms), and switching from stub-zone to forward-zone pointing to Consul). It looks like sometimes unbound was timing out returning a SERVFAIL making Kong retry with an A lookup instead.

It could be that Consul returns DNS records in a different fashion than your test server as I don't appear to be suffering from that.
